spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -444,85 +549,156 @@ public void run() {
+				future.get(this.pauseAfter, TimeUnit.MILLISECONDS);
+				if (this.logger.isDebugEnabled()) {
+					this.logger.debug("Success for " + record);
+				}
+				handleManualAcks();
+				this.listenerHasReturned = true;
+			}
+			catch (InterruptedException e) {
+				Thread.currentThread().interrupt();
+				throw new KafkaException("Listener thread interrupted", e);
+			}
+			catch (ExecutionException e) {
+				this.logger.debug("Failure for " + record + ": " + e.getMessage());
+				if (this.pauseWhenSlow && this.pauseException != null
+						&& this.pauseException.isAssignableFrom(e.getCause().getClass())) {
+					this.shouldPause = true;
+					this.listenerHasReturned = true;
+					this.pausedForException = true;
+				}
+				else {
+					if (getErrorHandler() != null) {
+						getErrorHandler().handle(e, record);
+					}
+					else {
+						this.logger.error("Listener threw an exception and no error handler for " + record, e);
+					}
+				}
+			}
+			catch (TimeoutException e) {
+				if (this.pauseWhenSlow) {
+					this.shouldPause = true;
+					future.addCallback(new ListenableFutureCallback<Void>() {
 
 						@Override
-						public void acknowledge() {
-							if (ListenerConsumer.this.ackMode.equals(AckMode.MANUAL)) {
-								updateManualOffset(record);
-							}
-							else if (ListenerConsumer.this.ackMode.equals(AckMode.MANUAL_IMMEDIATE)
-									|| ListenerConsumer.this.ackMode.equals(AckMode.MANUAL_IMMEDIATE_SYNC)) {
-								ackImmediate(record);
-							}
-							else {
-								throw new IllegalStateException("AckMode must be MANUAL, "
-										+ "MANUAL_IMMEDIATE, or MANUAL_IMMEDIATE_SYNC for manual acks");
+						public void onSuccess(Void result) {
+							if (ListenerConsumer.this.logger.isDebugEnabled()) {
+								ListenerConsumer.this.logger.debug("Delayed success for " + record);
 							}
+							ListenerConsumer.this.listenerHasReturned = true;
+							ListenerConsumer.this.consumer.wakeup();
 						}
 
-						private void ackImmediate(final ConsumerRecord<K, V> record) {
-							if (Thread.currentThread().equals(ListenerConsumer.this.consumerThread)) {
-								Map<TopicPartition, OffsetAndMetadata> commits = Collections.singletonMap(
-										new TopicPartition(record.topic(), record.partition()),
-										new OffsetAndMetadata(record.offset() + 1));
-								if (ListenerConsumer.this.logger.isDebugEnabled()) {
-									ListenerConsumer.this.logger.debug("Committing: " + commits);
-								}
-								if (ListenerConsumer.this.ackMode.equals(AckMode.MANUAL_IMMEDIATE)) {
-									ListenerConsumer.this.consumer.commitAsync(commits,
-											ListenerConsumer.this.commitCallback);
+						@Override
+						public void onFailure(Throwable ex) {
+							if (ListenerConsumer.this.logger.isDebugEnabled()) {
+								ListenerConsumer.this.logger.debug(
+										"Delayed failure for " + record + ": " + ex.getMessage());
+							}
+							if (ListenerConsumer.this.pauseWhenSlow
```

Comment:
`ListenerConsumer.this.pauseWhenSlow` is redundant here. We are within the `if (this.pauseWhenSlow) {` for the `final` property.